dataframely/_storage/delta.py

```diff
@@ -0,0 +1,201 @@
+        self, lf: pl.LazyFrame, serialized_schema: SerializedSchema, **kwargs: Any
+    ) -> None:
+        _raise_on_lazy_write()
+
+    def write_frame(
+        self, df: pl.DataFrame, serialized_schema: SerializedSchema, **kwargs: Any
+    ) -> None:
+        import deltalake
+
+        target = kwargs.pop("target")
+        metadata = kwargs.pop("metadata", {})
+        delta_write_options = kwargs.pop("delta_write_options", {})
+
+        # Delta lake does not allow partitioning if there is only one column
+        # We dynamically remove this setting here to allow users to still specify it
+        # on the collection level without having to worry about each individual member
+        if len(df.columns) < 2:
+            delta_write_options.pop("partition_by", None)
+
+        df.write_delta(
+            target,
+            delta_write_options=delta_write_options
+            | {
+                "commit_properties": deltalake.CommitProperties(
+                    custom_metadata=metadata | {SCHEMA_METADATA_KEY: serialized_schema}
+                ),
+            },
+            **kwargs,
+        )
+
+    def scan_frame(self, **kwargs: Any) -> tuple[pl.LazyFrame, SerializedSchema | None]:
+        table = _to_delta_table(kwargs.pop("source"))
+        serialized_schema = _read_serialized_schema(table)
+        df = pl.scan_delta(table, **kwargs)
+        return df, serialized_schema
+
+    def read_frame(self, **kwargs: Any) -> tuple[pl.DataFrame, SerializedSchema | None]:
+        table = _to_delta_table(kwargs.pop("source"))
+        serialized_schema = _read_serialized_schema(table)
+        df = pl.read_delta(table, **kwargs)
+        return df, serialized_schema
+
+    # ------------------------------ Collections ---------------------------------------
+    def sink_collection(
+        self,
+        dfs: dict[str, pl.LazyFrame],
+        serialized_collection: SerializedCollection,
+        serialized_schemas: dict[str, str],
+        **kwargs: Any,
+    ) -> None:
+        _raise_on_lazy_write()
+
+    def write_collection(
+        self,
+        dfs: dict[str, pl.LazyFrame],
+        serialized_collection: SerializedCollection,
+        serialized_schemas: dict[str, str],
+        **kwargs: Any,
+    ) -> None:
+        uri = Path(kwargs.pop("target"))
+
+        # The collection schema is serialized as part of the member parquet metadata
+        kwargs["metadata"] = kwargs.get("metadata", {}) | {
+            COLLECTION_METADATA_KEY: serialized_collection
+        }
+
+        for key, lf in dfs.items():
+            self.write_frame(
+                lf.collect(),
+                serialized_schema=serialized_schemas[key],
+                target=uri / key,
+                **kwargs,
+            )
+
+    def scan_collection(
+        self, members: Iterable[str], **kwargs: Any
+    ) -> tuple[dict[str, pl.LazyFrame], list[SerializedCollection | None]]:
+        import deltalake
+
+        uri = Path(kwargs.pop("source"))
```

Comment:
This could be an S3-path, which is not compatible with `Path`

Reply:
We _could_ use `universal_pathlib` with `fsspec` but that's quite the dependency tree to add

Reply:
Good point.

> that's quite the dependency tree to add

What do you mean by this? upath depends on fsspec, but fsspec has no required dependencies other than python, so it seems rather lightweight